lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,348 @@
+    """Exception thrown when an opensearch plugin misses installed dependencies."""
+
+    def __init__(self, name: str, missing_dependencies: List[str]):
+        self._deps = missing_dependencies
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install {self._name}, missing dependencies: {self._deps}"
+
+
+class OpenSearchPluginInstallError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin installation fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install plugin {self._name}: {self._msg}"
+
+
+class OpenSearchPluginRemoveError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin removal fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to remove plugin {self._name}: {self._msg}"
+
+
+class PluginState(BaseStrEnum):
+    """Enum for the states possible in plugins' lifecycle."""
+
+    MISSING = "missing"
+    AVAILABLE = "available"
+    ENABLED = "enabled"
+    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
+
+
+class OpenSearchPluginConfig:
+    """Represents the default configuration for a plugin."""
+
+    def __init__(
+        self,
+        config_entries: Optional[Dict[str, Any]] = None,
+        secret_entries: Optional[Dict[str, str]] = None,
+    ):
+        self.config_entries = config_entries
+        self.secret_entries = secret_entries
+
+
+class OpenSearchPlugin:
+    """Abstract class describing an OpenSearch plugin."""
+
+    PLUGIN_PROPERTIES = "plugin-descriptor.properties"
+
+    def __init__(self, plugins_path: str, extra_config: Dict[str, Any] = None):
+        """Creates the OpenSearchPlugin object.
+
+        Arguments:
+          plugins_path: str, path to the plugins folder
+          extra_config: dict, contains the data coming from either the config option
+                              or the relation bag
+        """
+        self._plugins_path = plugins_path
+        self._extra_config = extra_config
+        self._properties = Properties()
+
+    @property
+    def version(self) -> str:
+        """Returns the current version of the plugin.
+
+        Returns: str, string with the version code for this plugin
+        Raises:
+            FileNotFoundError: if plugin file is not present
+            PermissionError: if plugin file is present, but not set with correct permissions
+        """
+        plugin_props_path = os.path.join(f"{self._plugins_path}", f"{self.PLUGIN_PROPERTIES}")
+        with open(plugin_props_path) as f:
+            self._properties.load(f.read())
+        return self._properties._properties["version"]
```

Comment:
Given our charms run in linux, I think it's safe to use `/`:
```suggestion
        properties = Properties()
        with open(self._plugins_path) as f:
            properties.load(f.read())
        return properties._properties["version"]
```